#FUNCTIONS
1. What is the difference between a function and a method in Python?
 - A function is a block of code that is defined and can be called by name. It can exist independently and may or may not operate on an object.
 - A method is a function that is associated with an object and typically operates on that object (using self in case of class methods). It is bound to the instance of a class and can access and modify the instance's attributes.

2. Explain the concept of function arguments and parameters in Python.
 - Parameters are the variables that are defined in the function definition and specify what kind of values the function expects when called.
 - Arguments are the actual values or data passed to the function when it is called. They correspond to the function's parameters.

3. What are the different ways to define and call a function in Python?
 - A function is defined using the def keyword, followed by the function name and parentheses:
    def my_function(parameter):
    return parameter
 - The function is called by using its name followed by parentheses:
  my_function(value)

4. What is the purpose of the `return` statement in a Python function?
 - The return statement is used to send the result of a function back to the caller. When a function is called, it performs its operations and then uses return to pass the result out, effectively ending the function's execution and returning control to the caller.

5. What are iterators in Python and how do they differ from iterables?
 - An iterable is any Python object that can be iterated over (i.e., you can use a for loop on it). Examples include lists, tuples, and dictionaries.
 - An iterator is an object that represents a stream of data and has the methods __iter__() and __next__() to produce values one at a time. An iterator can be exhausted, meaning it can run out of values, whereas an iterable can be used to create multiple iterators.

6. Explain the concept of generators in Python and how they are defined.
 - Generators are a type of iterable that generate items lazily (i.e., one at a time and only when needed). They are defined using a function with the yield keyword instead of return. When yield is encountered, the function's state is saved, and it can be resumed later.
  - def my_generator():
    yield 1
    yield 2
    yield 3


7. What are the advantages of using generators over regular functions?
 - Memory Efficiency: Generators only produce one item at a time, rather than creating and storing an entire list in memory.
 - Lazy Evaluation: Values are computed only when requested, which can lead to faster execution when dealing with large datasets.
 - Infinite Sequences: Generators can be used to represent infinite sequences without consuming an infinite amount of memory.

8. What is a lambda function in Python and when is it typically used?
 - A lambda function is a small, anonymous function defined using the lambda keyword. It can take any number of arguments but can only have one expression. It is typically used for simple operations or when a short function is needed temporarily, such as in functional programming functions like map(), filter(), or sorted().
  - square = lambda x: x * x

9. Explain the purpose and usage of the `map()` function in Python.
 - The map() function applies a given function to all items in an iterable (such as a list) and returns an iterator yielding the results. It allows you to perform an operation on each element of a sequence in a concise way.
  - result = map(lambda x: x ** 2, [1, 2, 3, 4])  # [1, 4, 9, 16]

10. What is the difference between `map()`, `reduce()`, and `filter()` functions in Python?
 - map(): Applies a given function to all items in an iterable and returns an iterator with the results. Example: map(func, iterable).
 - reduce(): Applies a function of two arguments cumulatively to the items of an iterable, from left to right, so as to reduce the iterable to a single value. It is part of the functools module. Example: reduce(func, iterable).
 - filter(): Filters the items in an iterable by applying a function that returns True or False and returns an iterator containing only the items for which the function returned True. Example: filter(func, iterable).

 11. Using pen & Paper write the internal mechanism for sum operation using  reduce function on this given
 list:[47,11,42,13];
 - To understand how the reduce() function works for summing elements of a list, let's break it down step by step, using pen and paper (i.e., manually).

 - The given list is: [47, 11, 42, 13]

   Steps for Summing Using reduce() Function
   a) Import the reduce function from the functools module:
   - from functools import reduce
   b) Define the function that will be applied to the elements in the list. In this case, we want to sum the elements, so we can define a function as:
   - def add(x, y):
    return x + y
   c) Apply reduce() on the list with the add function:
   - result = reduce(add, [47, 11, 42, 13])

  Breakdown of the Process
  The reduce() function works by taking the first two elements of the list and applying the function, then taking the result of that and applying it with the next element, and so on. So, for our list [47, 11, 42, 13], here’s the step-by-step computation:

 Step 1: Apply add on the first two elements: 47 and 11.
 47 + 11 = 58

 Step 2: Take the result from Step 1 (58) and apply the add function to the next element (42).
 58 + 42 = 100

 Step 3: Take the result from Step 2 (100) and apply the add function to the next element (13).
 100 + 13 = 113

- Final Result
 - After all the steps, the final result is 113.
  So, using reduce(), the sum of the list [47, 11, 42, 13] is 113.


 #PRACTICALS QUESTIONS
 1. Write a Python function that takes a list of numbers as input and returns the sum of all even numbers in
the list.






In [4]:
def sum_even_numbers(numbers):
    return sum(num for num in numbers if num % 2 == 0)

# Example usage:
numbers = [1, 2, 3, 4, 5, 6]
result = sum_even_numbers(numbers)
print(result)


12


2. Create a Python function that accepts a string and returns the reverse of that string.

In [5]:
def reverse_string(s):
    return s[::-1]

# Example usage:
input_string = "Hello, World!"
result = reverse_string(input_string)
print(result)


!dlroW ,olleH


3. Implement a Python function that takes a list of integers and returns a new list containing the squares of each number.

In [6]:
def square_numbers(numbers):
    return [num ** 2 for num in numbers]

# Example usage:
numbers = [1, 2, 3, 4]
result = square_numbers(numbers)
print(result)


[1, 4, 9, 16]


4. Python Function to Check if a Number is Prime (1 to 200)


In [8]:
def is_prime(num):
    if num <= 1:
        return False
    for i in range(2, int(num ** 0.5) + 1):
        if num % i == 0:
            return False
    return True

# Example usage:
primes = [num for num in range(1, 201) if is_prime(num)]
print(primes)


[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199]


5. Create an iterator class in Python that generates the Fibonacci sequence up to a specified number of terms.


In [9]:
class FibonacciIterator:
    def __init__(self, n):
        self.n = n
        self.a, self.b = 0, 1
        self.count = 0

    def __iter__(self):
        return self

    def __next__(self):
        if self.count < self.n:
            result = self.a
            self.a, self.b = self.b, self.a + self.b
            self.count += 1
            return result
        else:
            raise StopIteration

# Example usage:
fibonacci = FibonacciIterator(10)
for num in fibonacci:
    print(num)


0
1
1
2
3
5
8
13
21
34


6. Write a generator function in Python that yields the powers of 2 up to a given exponent.

In [10]:
def powers_of_2(exponent):
    for i in range(exponent + 1):
        yield 2 ** i

# Example usage:
for power in powers_of_2(5):
    print(power)


1
2
4
8
16
32


8. Use a lambda function in Python to sort a list of tuples based on the second element of each tuple.

In [17]:
tuples = [(1, 'apple'), (2, 'banana'), (3, 'cherry')]
sorted_tuples = sorted(tuples, key=lambda x: x[1])

# Example usage:
print(sorted_tuples)


[(1, 'apple'), (2, 'banana'), (3, 'cherry')]


9. Write a Python program that uses `map()` to convert a list of temperatures from Celsius to Fahrenheit.

In [18]:
def celsius_to_fahrenheit(celsius):
    return (celsius * 9/5) + 32

# Example usage:
temperatures_celsius = [0, 20, 37, 100]
temperatures_fahrenheit = list(map(celsius_to_fahrenheit, temperatures_celsius))
print(temperatures_fahrenheit)


[32.0, 68.0, 98.6, 212.0]


10. Create a Python program that uses `filter()` to remove all the vowels from a given string.




In [19]:
def remove_vowels(s):
    vowels = "aeiouAEIOU"
    return ''.join(filter(lambda x: x not in vowels, s))

# Example usage:
input_string = "Hello, World!"
result = remove_vowels(input_string)
print(result)


Hll, Wrld!


11) Imagine an accounting routine used in a book shop. It works on a list with sublists, which look like this:
 order number- 34587,
               98762,
               77226,
               88112
 book title and author- learning python,
                        Mark Lutz
                    programming python, Mark Lutz
                    Head first python, Paul Barry
                    Einfuhrung in python3, Bernd klein
 quantity - 4,5,3,3
 price - 40.95, 56.80, 32.95, 24.99

 - Write a Python program, which returns a list with 2-tuples. Each tuple consists of the order number and the
product of the price per item and the quantity. The product should be increased by 10,- € if the value of the
order is smaller than 100,00 €.

 - Write a Python program using lambda and map.

 ans- To implement the accounting routine as described, we need to:

 1)Calculate the total value of each order by multiplying the price per item and the quantity.
 2)If the total value is less than 100€, increase the value by 10€.
 3)Return the result as a list of 2-tuples, where each tuple consists of the order number and the adjusted total price.
  Additionally, we'll use lambda and map() to solve the problem in a functional programming style.
  

In [20]:
# Data: order numbers, book titles/authors, quantities, prices
order_numbers = [34587, 98762, 77226, 88112]
book_titles_authors = [
    ("Learning Python", "Mark Lutz"),
    ("Programming Python", "Mark Lutz"),
    ("Head First Python", "Paul Barry"),
    ("Einfuhrung in Python3", "Bernd Klein")
]
quantities = [4, 5, 3, 3]
prices = [40.95, 56.80, 32.95, 24.99]

# Lambda function to calculate the total price per order and apply the adjustment if necessary
calculate_order_value = lambda order, qty, price: (order, qty * price + 10 if qty * price < 100 else qty * price)

# Using map to apply the function on each order
order_values = list(map(lambda order, qty, price: calculate_order_value(order, qty, price), order_numbers, quantities, prices))

# Print the result
print(order_values)


[(34587, 163.8), (98762, 284.0), (77226, 108.85000000000001), (88112, 84.97)]


Breakdown of the Code:
- Data: We have four lists, which contain:
order_numbers: Order IDs
book_titles_authors: Titles and authors of books (not used in the calculation, but kept for clarity).
quantities: The number of books sold in each order.
prices: The price per book.

- Lambda Function:
 - We define a lambda function calculate_order_value that takes three arguments: order, qty, and price.
 - It calculates the total value for each order (qty * price).
 - If the total value is less than 100€, we add 10€ to it.

 - Map:
  - We use the map() function to apply the lambda function to each corresponding element of order_numbers, quantities, and prices.
  - The lambda function calls calculate_order_value to compute the adjusted order value.

- Result:
  We convert the output of map() to a list and print it. Each tuple contains the order number and the calculated order value, with any necessary adjustments applied.



In [21]:
[
    (34587, 163.8),  # 40.95 * 4 = 163.8
    (98762, 284.0),  # 56.80 * 5 = 284.0
    (77226, 98.85),  # 32.95 * 3 = 98.85
    (88112, 74.97)   # 24.99 * 3 = 74.97 (increased by 10 because it's less than 100)
]


[(34587, 163.8), (98762, 284.0), (77226, 98.85), (88112, 74.97)]

- In this output:
Orders with totals above 100€ remain unchanged.
The order with a total of 74.97€ is adjusted by adding 10€, resulting in 84.97€.
This program uses lambda and map() to calculate and adjust the values for each order in a concise, functional style.

7. Implement a generator function that reads a file line by line and yields each line as a string.

In [26]:
def read_file_line_by_line(file_path):
    with open(file_path, 'r') as file:
        for line in file:
            yield line

